src/site/topic2-cpp.rst

```diff
@@ -0,0 +1,648 @@
+        System.out.println("it was true");
+    } else {
+        System.out.println("it was false");
+    }
+
+
+**C++**
+
+.. code-block:: cpp
+    :linenos:
+
+    // C++ --- if/else
+    // Assume someBool exists and is a boolean
+    if (someBool) {
+        std::cout << "it was true" << std::endl;
+    } else {
+	    std::cout << "it was false" << std::endl;
+    }
+
+
+Boolean Operators
+-----------------
+
+* Same as Java
+    * less than --- ``a < b``
+    * sameness --- ``c == d``
+    * *not* sameness --- ``e != f``
+
+* We've also got our logical operators
+    * and --- ``v && w``
+    * or --- ``x || y``
+    * not --- ``!z``
+
+
+Loops
+=====
+
+While Loops
+-----------
+
+**Python**
+
+.. code-block:: Python
+    :linenos:
+
+    # Python --- while loop
+    c = 0
+
+    # While some condition is true
+    while c < 10:
+        print("c is now: " + str(c))
+        c+=1
+
+**Java**
+
+.. code-block:: Java
+    :linenos:
+    :emphasize-lines: 5
+
+    // Java --- while loop
+    int c = 0;
+
+    // While some condition is true
+    while (c < 10) {
+        System.out.println("c is now: " + c);
+        c++;
+    }
+
+
+**C++**
+
+.. code-block:: cpp
+    :linenos:
+    :emphasize-lines: 5
+
+    // C++ --- while loop
+    int c = 0;
+
+    // While some condition is true
+    while (c < 10) {
+        std::cout << "c is now: " << c << std::endl;
+        c++;
+    }
+
+
+More While Loops
+----------------
+
+**Python**
+
+.. code-block:: Python
+    :linenos:
+
+    # Python --- while loop again
+    stop = False
+    c = 0
+
+    while not stop:
+        print("c is now: " + str(c))
+        c+=1
+        if c == 5:
+            stop = True
+
+**Java**
+
+.. code-block:: Java
+    :linenos:
+
+    // Java --- while loop again
+    boolean stop = false;
+    int c = 0;
+
+    while (!stop) {
+        System.out.println("c is now: " + c);
+        c++;
+        if (c == 5) {
+            stop = true;
+        }
+    }
+
+
+**C++**
+
+.. code-block:: cpp
+    :linenos:
+
+    // C++ --- while loop again
+    bool stop = false;
+    int c = 0;
+
+    while (!stop) {
+        std::cout << "c is now: " << c << std::endl;
+        c++;
+        if (c == 5) {
+            stop = true;
+        }
+    }
+
+
+For Each Loop
+-------------
+
+**Python**
+
+.. code-block:: Python
+    :linenos:
+
+    # Python --- for loop
+    aList = ['a', 'b', 'c', 'd']
+
+    # For each thing 'c' in aList
+    for c in aList:
+        print(c)
+
+**Java**
+
+.. code-block:: Java
+    :linenos:
+
+    // Java --- for each loop
+    char[] anArray = {'a', 'b', 'c', 'd'};
+
+    // For each character 'c' in anArray
+    for (char e : anArray) {
+        System.out.println(e);
+    }
+
+
+**C++**
+
+.. code-block:: cpp
+    :linenos:
+    :emphasize-lines: 2
+
+    // C++ --- for each loop
+    char anArray[] = {'a', 'b', 'c', 'd'};
+
+    // For each character 'c' in anArray
+    for (char e : anArray) {
+        std::cout << e << std::endl;
+    }
```

Comment:
Oddly enough, even though this is c++11, I've never seen this before. I've always seen stuff like:
```cpp
for (auto it = v.begin(); it != v.end(); ++it) {}
```

Reply:
I'll add that too, but, another great example of how terrible this is. 

Students "wtf is an _iterator?_"
Me "uhhh, come back in 2 months"

Reply:
Imagine what it was like before `auto`

Reply:
After thinking on this more --- I think deliberately _not_ going into this one. Hell, I find it confusing, and I worry it would just ruin `for` loops in both cpp and java for em'.

Reply:
I mean. Ranges are a pretty big deal in cpp. The for-each loop there is a [range based for loop](https://en.cppreference.com/w/cpp/language/range-for), you kinda need to know what a range is to use it.

You could make the same argument for `Iterable` in java, but at least java classes actually implement the interface. Ranges in cpp are just things with a meaningful return from `std::begin()` and `std::end()`. (Which have a `decltype` return type. Fun!)

Reply:
I hate every bit of this.

Reply:
> (Which have a decltype return type. Fun!)

False